core/chain_makers.go

```diff
@@ -318,3 +320,171 @@ func (cr *fakeChainReader) GetHeaderByNumber(number uint64) *types.Header
+	chainreader := &fakeChainReader{config: config}
+
+	resultLoop := func(i int, k *knot, results chan *types.Header, stop chan struct{}, wg *sync.WaitGroup) error {
+		for {
+			select {
+			case header := <-results:
+				k.block = types.NewBlockWithHeader(header)
+				fmt.Println("Mined block: ", k.block.Hash(), k.block.Header().NumberArray(), k.block.Header().DifficultyArray())
+				defer wg.Done()
+				return nil
+			}
+		}
+	}
+
+	genblock := func(i int, parent *types.Block, genesis *types.Header, primedb *state.StateDB, regiondb *state.StateDB, zonedb *state.StateDB, location common.Location, resultCh chan *types.Header, stopCh chan struct{}) {
+		b := &BlockGen{i: i, chain: blocks, parent: parent, statedb: primedb, config: config, engine: engine}
+		b.header = makeKnotHeader(chainreader, parent, genesis, primedb, b.engine, location, i)
+
+		// Execute any user modifications to the block
+		if gen != nil {
+			gen(i, b)
+		}
+		if b.engine != nil {
+			// Finalize and seal the block
+			b.engine.FinalizeAtIndex(chainreader, b.header, zonedb, b.txs, b.uncles, common.ZONE_CTX)
+			// Write state changes to db
+			root, err := zonedb.Commit(config.IsEIP158(b.header.Number(common.ZONE_CTX)))
+			if err != nil {
+				panic(fmt.Sprintf("state write error: %v", err))
+			}
+			if err := zonedb.Database().TrieDB().Commit(root, false, nil); err != nil {
+				panic(fmt.Sprintf("trie write error: %v", err))
+			}
+
+			b.engine.FinalizeAtIndex(chainreader, b.header, regiondb, b.txs, b.uncles, common.REGION_CTX)
+			// Write state changes to db
+			root, err = regiondb.Commit(config.IsEIP158(b.header.Number(common.REGION_CTX)))
+			if err != nil {
+				panic(fmt.Sprintf("state write error: %v", err))
+			}
+			if err := regiondb.Database().TrieDB().Commit(root, false, nil); err != nil {
+				panic(fmt.Sprintf("trie write error: %v", err))
+			}
+
+			block, _ := b.engine.FinalizeAndAssemble(chainreader, b.header, primedb, b.txs, b.uncles, b.receipts)
+
+			// Write state changes to db
+			root, err = primedb.Commit(config.IsEIP158(b.header.Number(common.PRIME_CTX)))
+			if err != nil {
+				panic(fmt.Sprintf("state write error: %v", err))
+			}
+			if err := primedb.Database().TrieDB().Commit(root, false, nil); err != nil {
+				panic(fmt.Sprintf("trie write error: %v", err))
+			}
+
+			// Mine the block
+			if err := engine.Seal(block.Header(), resultCh, stopCh); err != nil {
+				log.Println("Block sealing failed", "err", err)
+			}
+		}
+	}
+
+	knot := &knot{
+		block: parent,
+	}
+
+	genesis := parent.Header()
+	fmt.Println("genesis hash", genesis.Hash())
+
+	locations := []common.Location{{0, 0}, {0, 1}, {0, 2}, {1, 0}, {1, 1}, {1, 2}, {2, 0}, {2, 1}, {2, 2}}
+	for i := 0; i < len(locations); i++ {
+		primedb, err := state.New(parent.Root(), state.NewDatabase(db), nil)
+		if err != nil {
+			panic(err)
+		}
+
+		regionParentRoot := common.Hash{}
+		// mod 3 is amount of regions
+		if i%3 == 0 {
```

Comment:
Again, I'd prefer to avoid magic numbers. Perhaps `common.NumRegionsInPrime` or `common.NumZonesInRegion`?